src/Chapter4.hs

```diff
@@ -472,10 +477,10 @@ Implement the Applicative instance for our 'Secret' data type from before.
 -}
 instance Applicative (Secret e) where
     pure :: a -> Secret e a
-    pure = error "pure Secret: Not implemented!"
+    pure = Reward
 
     (<*>) :: Secret e (a -> b) -> Secret e a -> Secret e b
-    (<*>) = error "(<*>) Secret: Not implemented!"
+    (<*>) (Reward g) (Reward x) = Reward (g x) 
```

Comment:
I see that pattern-matching in this function is not exhaustive. For instance, this function doesn't handle cases when one of the arguments is `Trap`

Reply:
This is embarrassing. I somehow didn't push my local commit. Sorry about that 😞 . I've pushed the commit now.
Also, below is the code related to `Secret e` datatype - Functor, Applicative, and Monad instances.

``` haskell

instance Functor (Secret e) where
    fmap :: (a -> b) -> Secret e a -> Secret e b
    fmap g (Reward x) = Reward (g x)
    fmap _ (Trap x) = Trap x 


instance Applicative (Secret e) where
    pure :: a -> Secret e a
    pure = Reward

    (<*>) :: Secret e (a -> b) -> Secret e a -> Secret e b
    (<*>) (Reward g) (Reward x) = Reward (g x) 
    (<*>) (Trap x) _ = Trap x
    (<*>) _ (Trap x) = Trap x


instance Monad (Secret e) where
    (>>=) :: Secret e a -> (a -> Secret e b) -> Secret e b
    (>>=) (Reward x) g = g x
    (>>=) (Trap x) _ = Trap x
```

Reply:
@jhashekhar Hey, don't worry! The instances look good now :slightly_smiling_face: